spring-data-relational/src/main/java/org/springframework/data/relational/repository/query/RelationalExampleMapper.java

```diff
@@ -64,92 +72,194 @@ public <T> Query getMappedExample(Example<T> example) {
+					.apply(Optional.ofNullable(actualPropertyValue));
 
 			// If the value is empty, don't try to match against it
-			if (!optionalConvertedPropValue.isPresent()) {
+			if (optionalConvertedPropValue.isEmpty()) {
 				return;
 			}
 
 			Object convPropValue = optionalConvertedPropValue.get();
-			boolean ignoreCase = matcherAccessor.isIgnoreCaseForPath(property.getName());
+			boolean ignoreCase = matcherAccessor.isIgnoreCaseForPath(currentPropertyDotPath);
 
 			String column = property.getName();
 
-			switch (matcherAccessor.getStringMatcherForPath(property.getName())) {
+			switch (matcherAccessor.getStringMatcherForPath(currentPropertyDotPath)) {
 				case DEFAULT:
 				case EXACT:
-					criteriaBasedOnProperties.add(includeNulls(example) //
+					criteriaBasedOnProperties.add(includeNulls(matcherAccessor) //
 							? Criteria.where(column).isNull().or(column).is(convPropValue).ignoreCase(ignoreCase)
 							: Criteria.where(column).is(convPropValue).ignoreCase(ignoreCase));
 					break;
 				case ENDING:
-					criteriaBasedOnProperties.add(includeNulls(example) //
+					criteriaBasedOnProperties.add(includeNulls(matcherAccessor) //
 							? Criteria.where(column).isNull().or(column).like("%" + convPropValue).ignoreCase(ignoreCase)
 							: Criteria.where(column).like("%" + convPropValue).ignoreCase(ignoreCase));
 					break;
 				case STARTING:
-					criteriaBasedOnProperties.add(includeNulls(example) //
+					criteriaBasedOnProperties.add(includeNulls(matcherAccessor) //
 							? Criteria.where(column).isNull().or(column).like(convPropValue + "%").ignoreCase(ignoreCase)
 							: Criteria.where(column).like(convPropValue + "%").ignoreCase(ignoreCase));
 					break;
 				case CONTAINING:
-					criteriaBasedOnProperties.add(includeNulls(example) //
+					criteriaBasedOnProperties.add(includeNulls(matcherAccessor) //
 							? Criteria.where(column).isNull().or(column).like("%" + convPropValue + "%").ignoreCase(ignoreCase)
 							: Criteria.where(column).like("%" + convPropValue + "%").ignoreCase(ignoreCase));
 					break;
 				default:
-					throw new IllegalStateException(example.getMatcher().getDefaultStringMatcher() + " is not supported");
+					throw new IllegalStateException(matcherAccessor.getDefaultStringMatcher() + " is not supported");
 			}
-		});
+		}
 
-		// Criteria, assemble!
-		Criteria criteria = Criteria.empty();
+	}
 
-		for (Criteria propertyCriteria : criteriaBasedOnProperties) {
+	/**
+	 * Processes an embedded entity's properties recursively.
+	 *
+	 * @param matcherAccessor the input matcher on the {@link Example#getProbe() original probe}.
+	 * @param value the actual embedded object.
+	 * @param property the embedded property.
+	 * @param criteriaBasedOnProperties collection of {@link Criteria} objects to potentially enrich.
+	 * @param currentPropertyPath the dot-separated path of the passed {@code property}.
+	 */
+	private void processEmbeddedRecursively(
```

Comment:
Looks like a duplication of the general code that collects criteria for example matching. Should be consolidated.